File: merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        
        if (lists == null){
            return null;
        }
        
        /**
        
        ListNode head = null;
        ListNode cur = null, prevC = null;
        ListNode newNode;
        for (ListNode temp : lists){
            if(head == null){
                head = new ListNode(0);
                head.next = temp;
                cur = null;
                temp = null;
            }
            else{
                prevC = head;
                cur = head.next;
            }
            while(temp != null){
                while(cur != null && cur.val <= temp.val){
                    prevC = cur;
                    cur = cur.next;
                }
                prevC.next = temp;
                temp = temp.next;
                prevC.next.next = cur;
                prevC = prevC.next;
            }
        }
        return head == null ? null : head.next;
        
        **/
        
        PriorityQueue<ListNode> queue = new PriorityQueue<>((n1, n2) -> n1.val - n2.val);
        for (ListNode node : lists){
            while(node != null){
                queue.add(node);
                node = node.next;
            }
        }
        ListNode res = new ListNode(0);
        ListNode cur = res;
        while(!queue.isEmpty()){
            cur.next = queue.poll();
            cur = cur.next;
        }
        cur.next = null;
        return res.next;
    }
}

File: merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size() == 1){
            return lists[0];
        }
        if(lists.size() == 0){
            return NULL;
        }
        int j = 0;
        int i = 0;
        int len = lists.size() - 1;
        
        while(len != 0){
            j = len;
            i = 0;
            while(i < j){
                lists[i] = merge(lists[i], lists[j]);
                j--;
                i++;
                if(i >= j){
                    len = j;
                }
            }
        }
        
        return lists[0];
    }
    
    
    ListNode * merge(ListNode* left, ListNode* right){
        ListNode* head = new ListNode(0);
        ListNode* cur = head;
        
        while(left != NULL && right != NULL){
            if(left->val > right->val){
                cur->next = right;
                cur = right;
                right = right->next;
                cur->next = NULL;
            }
            else{
                cur->next = left;
                cur = left;
                left = left->next;
                cur->next = NULL;
            }
        }
        
        if(left != NULL){
            cur->next = left;
        }
        
        if(right != NULL){
            cur->next = right;
        }
        
        return head->next;
    }
};